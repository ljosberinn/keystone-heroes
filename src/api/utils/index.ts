import { Role } from "@prisma/client";

export const sortByRole = (a: Role, b: Role): 0 | -1 | 1 => {
  if (b === Role.tank || a === Role.dps) {
    return 1;
  }

  if (b === Role.dps || a === Role.tank) {
    return -1;
  }

  return 0;
};
