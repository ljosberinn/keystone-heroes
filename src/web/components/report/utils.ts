import type { FightSuccessResponse } from "../../../api/functions/fight";
import {
  SANGUINE_ICHOR_HEALING,
  spells,
  PF,
  EXPLOSIVE,
} from "../../staticData";
import type { AbilityReadyRowProps } from "./rows/AbilityReadyRow";
import type { ApplyBuffRowProps } from "./rows/ApplyBuffRow";
import type { ApplyDebuffRowProps } from "./rows/ApplyDebuffRow";
import type { CastRowProps } from "./rows/CastRow";
import type { DamageDoneRowProps } from "./rows/DamageDoneRow";
import type { DamageTakenRowProps } from "./rows/DamageTakenRow";
import type { DeathRowProps } from "./rows/DeathRow";
import type { ExplosiveSummaryRowProps } from "./rows/ExplosivesSummaryRow";
import type { HealingDoneRowProps } from "./rows/HealingDoneRow";
import type { InterruptRowProps } from "./rows/InterruptRow";
import type { MissedInterruptRowProps } from "./rows/MissedInterruptRow";
import type { PlagueBombDamageRowProps } from "./rows/PlagueBombDamageRow";
import type { SanguineTimeLossRowProps } from "./rows/SanguineTimeLossRow";
import type { ViolentDetonationDamageRowProps } from "./rows/ViolentDetonationDamageRow";

const bloodlustTypes = new Set([
  // Heroism
  32_182,
  // Drums of Deathly Ferocity
  309_658,
  // Timewarp
  80_353,
  // Bloodlust
  2825,
]);
const invisibilityTypes = new Set([
  // Invisible; Potion of the Hidden Spirit
  307_195,
  // Dimensional Shifter
  321_422,
]);

export const findBloodlust = <
  T extends { events: FightSuccessResponse["pulls"][number]["events"] }
>(
  pull: T
): number | null => {
  const match = pull.events.find(
    (event) =>
      event.type !== "AbilityReady" &&
      event.ability &&
      bloodlustTypes.has(event.ability.id)
  );

  return match?.ability?.id ?? null;
};

export const detectInvisibilityUsage = (
  pull: FightSuccessResponse["pulls"][number]
): null | "invisibility" | "shroud" => {
  const invisEvent = pull.events.find(
    (event) =>
      event.type === "ApplyBuff" &&
      event.ability &&
      invisibilityTypes.has(event.ability.id) &&
      event.category === "AFTER"
  );

  if (invisEvent) {
    return "invisibility";
  }

  const shroudEvent = pull.events.find(
    (event) =>
      event.type === "Cast" &&
      event.ability?.id === 114_018 &&
      event.category === "AFTER"
  );

  return shroudEvent ? "shroud" : null;
};

export const calcChallengersBurden = (keyLevel: number): number => {
  return Math.round((1.08 ** (keyLevel - 2) - 1) * 100);
};

export const determineAbility = (id: number): typeof spells[number] | null => {
  if (spells[id]) {
    return spells[id];
  }

  // Engineer rez
  if (id === 345_130) {
    return {
      name: "Disposable Spectrophasic Reanimator",
      icon: "inv_engineering_90_lightningbox",
      cd: 0,
    };
  }

  return null;
};

export type DefaultEvent =
  FightSuccessResponse["pulls"][number]["events"][number];

export const isCastEventWithAbilityAndSourcePlayer = (
  event: DefaultEvent
): event is CastRowProps["event"] =>
  (event.type === "Cast" || event.type === "BeginCast") &&
  event.ability !== null &&
  event.sourcePlayerID !== null;

export const isAbilityReadyEventWithAbilityAndSourcePlayer = (
  event: DefaultEvent
): event is AbilityReadyRowProps["event"] =>
  event.type === "AbilityReady" &&
  event.ability !== null &&
  event.sourcePlayerID !== null;

export const isDamageTakenEventWithTargetPlayer = (
  event: DefaultEvent
): event is DamageTakenRowProps["event"] =>
  event.type === "DamageTaken" &&
  event.targetPlayerID !== null &&
  event.damage !== null;

export const isInterruptEventWithSourceAndTargetPlayerAndAbility = (
  event: DefaultEvent
): event is InterruptRowProps["event"] =>
  event.type === "Interrupt" &&
  event.sourcePlayerID !== null &&
  event.ability !== null &&
  event.targetPlayerID !== null;

export type PlayerDeathEvent = Omit<DefaultEvent, "type"> & {
  type: "Death";
  sourceNPC: DefaultEvent["sourceNPC"];
  targetPlayerID: NonNullable<DefaultEvent["targetPlayerID"]>;
};

export type NPCDeathEvent = Omit<DefaultEvent, "type"> & {
  type: "Death";
  targetNPC: NonNullable<DefaultEvent["targetNPC"]>;
};

const isNPCDeathEvent = (event: DefaultEvent): event is NPCDeathEvent =>
  event.type === "Death" && event.targetNPC !== null;
const isPlayerDeathEvent = (event: DefaultEvent): event is PlayerDeathEvent =>
  event.type === "Death" && event.targetPlayerID !== null;

export const isPlayerOrNPCDeathEvent = (
  event: DefaultEvent
): event is DeathRowProps["event"] => {
  return isNPCDeathEvent(event) || isPlayerDeathEvent(event);
};

export const isDamageDoneEventWithAbility = (
  event: DefaultEvent
): event is DamageDoneRowProps["event"] =>
  event.type === "DamageDone" &&
  event.ability !== null &&
  event.damage !== null &&
  event.sourcePlayerID !== null;

export const isHealingDoneEventWithAbility = (
  event: DefaultEvent
): event is HealingDoneRowProps["event"] =>
  event.type === "HealingDone" &&
  event.ability !== null &&
  ("sourcePlayerID" in event || "targetNPC" in event);

export const isApplyBuffEventWithAbility = (
  event: DefaultEvent
): event is ApplyBuffRowProps["event"] =>
  (event.type === "ApplyBuff" ||
    event.type === "ApplyBuffStack" ||
    event.type === "RemoveBuff") &&
  event.ability !== null;

export const isApplyDebuffEventWithAbility = (
  event: DefaultEvent
): event is ApplyDebuffRowProps["event"] =>
  (event.type === "ApplyDebuff" || event.type === "ApplyDebuffStack") &&
  event.ability !== null;

export const isSanguineHealEvent = (
  event: DefaultEvent
): event is SanguineTimeLossRowProps["events"][number] =>
  event.type === "HealingDone" && event.ability?.id === SANGUINE_ICHOR_HEALING;

export const isPlagueBombDamageEvent = (
  event: DefaultEvent
): event is PlagueBombDamageRowProps["events"][number] =>
  (event.type === "DamageDone" || event.type === "DamageTaken") &&
  event.ability?.id === PF.PLAGUE_BOMB;

export const isViolentDetonationDamageEvent = (
  event: DefaultEvent
): event is ViolentDetonationDamageRowProps["events"][number] =>
  (event.type === "DamageDone" || event.type === "DamageTaken") &&
  event.ability?.id === PF.CANISTER_VIOLENT_DETONATION;

export const isExplosivesDamageEvent = (
  event: DefaultEvent
): event is ExplosiveSummaryRowProps["events"][number] =>
  event.type === "DamageDone" && event.targetNPC?.id === EXPLOSIVE.unit;

export const isMissingInterruptEventWithAbility = (
  event: DefaultEvent
): event is MissedInterruptRowProps["event"] =>
  event.type === "MissedInterrupt" && event.ability !== null;

export const formatNumber = (value: number) => {
  if (value > 1000000) {
    return `${Number.parseFloat((value / 1000000).toFixed(2)).toLocaleString(
      "en-US"
    )}m`;
  }

  if (value > 1000) {
    return `${Number.parseFloat((value / 1000).toFixed(2)).toLocaleString(
      "en-US"
    )}k`;
  }

  return value.toLocaleString("en-US");
};
