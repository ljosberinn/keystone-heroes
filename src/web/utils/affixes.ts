import type { FightSuccessResponse } from "../../api/functions/fight";
import {
  BURSTING,
  ENCRYPTED,
  EXPLOSIVE,
  GRIEVOUS,
  NECROTIC,
  QUAKING,
  SANGUINE_ICHOR_DAMAGE,
  SANGUINE_ICHOR_HEALING,
  SPITEFUL,
  STORMING,
  VOLCANIC,
} from "../staticData";

export type CalculateAffixMetricsParams = {
  events: FightSuccessResponse["pulls"][number]["events"];
  affixes: FightSuccessResponse["affixes"];
  groupDPS: FightSuccessResponse["meta"]["dps"];
};

export type SanguineMetrics = {
  hasSanguine: boolean;
  healing: number;
  estTimeLoss: number;
  damage: Record<number, number>;
  hits: Record<number, number>;
};

export const calculateSanguineMetrics = ({
  affixes,
  events,
  groupDPS,
}: CalculateAffixMetricsParams): SanguineMetrics => {
  if (!affixes.includes(8)) {
    return {
      healing: 0,
      damage: {},
      hasSanguine: false,
      estTimeLoss: 0,
      hits: {},
    };
  }

  const healing = events.reduce((acc, event) => {
    if (
      event.type === "HealingDone" &&
      event.healingDone &&
      event.ability?.id === SANGUINE_ICHOR_HEALING
    ) {
      return acc + event.healingDone;
    }

    return acc;
  }, 0);

  const { hits, damage } = events.reduce<{
    hits: Record<number, number>;
    damage: Record<number, number>;
  }>(
    (acc, event) => {
      if (
        event.type === "DamageTaken" &&
        event.ability?.id === SANGUINE_ICHOR_DAMAGE &&
        event.targetPlayerID
      ) {
        acc.hits[event.targetPlayerID] =
          (acc.hits[event.targetPlayerID] ?? 0) + 1;

        if (event.damage) {
          acc.damage[event.targetPlayerID] =
            (acc.damage[event.targetPlayerID] ?? 0) + event.damage;
        }
      }

      return acc;
    },
    {
      hits: {},
      damage: {},
    }
  );

  return {
    hasSanguine: true,
    healing,
    damage,
    estTimeLoss: Math.floor(healing / groupDPS),
    hits,
  };
};

export type GrievousMetrics = {
  hasGrievous: boolean;
  damage: Record<number, number>;
};

export const calculateGrievousMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): GrievousMetrics => {
  if (!affixes.includes(12)) {
    return {
      damage: {},
      hasGrievous: false,
    };
  }

  const damage = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageTaken" &&
      event.ability?.id === GRIEVOUS &&
      event.targetPlayerID &&
      event.damage
    ) {
      acc[event.targetPlayerID] =
        (acc[event.targetPlayerID] ?? 0) + event.damage;
    }
    return acc;
  }, {});

  return {
    hasGrievous: true,
    damage,
  };
};

export type BurstingMetrics = {
  hasBursting: boolean;
  damage: Record<number, number>;
};

export const calculateBurstingMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): BurstingMetrics => {
  if (!affixes.includes(11)) {
    return {
      damage: {},
      hasBursting: false,
    };
  }

  const damage = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageTaken" &&
      event.ability?.id === BURSTING.damage &&
      event.targetPlayerID &&
      event.damage
    ) {
      acc[event.targetPlayerID] =
        (acc[event.targetPlayerID] ?? 0) + event.damage;
    }
    return acc;
  }, {});

  return {
    hasBursting: true,
    damage,
  };
};

export type SpitefulMetrics = {
  hasSpiteful: boolean;
  damage: Record<number, number>;
  hits: Record<number, number>;
};

export const calculateSpitefulMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): SpitefulMetrics => {
  if (!affixes.includes(123)) {
    return {
      damage: {},
      hits: {},
      hasSpiteful: false,
    };
  }

  const { hits, damage } = events.reduce<{
    hits: Record<number, number>;
    damage: Record<number, number>;
  }>(
    (acc, event) => {
      if (
        event.type === "DamageTaken" &&
        event.ability?.id === SPITEFUL &&
        event.targetPlayerID
      ) {
        acc.hits[event.targetPlayerID] =
          (acc.hits[event.targetPlayerID] ?? 0) + 1;

        if (event.damage) {
          acc.damage[event.targetPlayerID] =
            (acc.damage[event.targetPlayerID] ?? 0) + event.damage;
        }
      }

      return acc;
    },
    {
      hits: {},
      damage: {},
    }
  );

  return {
    damage,
    hits,
    hasSpiteful: true,
  };
};

export type ExplosiveMetrics = {
  hasExplosive: boolean;
  kills: Record<number, number>;
  spawned: number;
  missed: number;
};

export const calculateExplosiveMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): ExplosiveMetrics => {
  if (!affixes.includes(13)) {
    return {
      hasExplosive: false,
      spawned: 0,
      kills: {},
      missed: 0,
    };
  }

  const kills = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageDone" &&
      event.targetNPC?.id === EXPLOSIVE.unit &&
      event.sourcePlayerID
    ) {
      acc[event.sourcePlayerID] = (acc[event.sourcePlayerID] ?? 0) + 1;
    }

    return acc;
  }, {});

  const missed = Object.values(
    events.reduce<Record<number, number>>((acc, event) => {
      if (
        event.type === "DamageTaken" &&
        event.ability?.id === EXPLOSIVE.ability
      ) {
        acc[event.timestamp] = (acc[event.timestamp] ?? 0) + 1;
      }

      return acc;
    }, {})
  ).length;

  const spawned =
    Object.values(kills).reduce((acc, count) => acc + count, 0) + missed;

  return {
    hasExplosive: true,
    kills,
    spawned,
    missed,
  };
};

export type QuakingMetrics = {
  hasQuaking: boolean;
  damage: Record<number, number>;
  interrupts: Record<number, number>;
};

export const calculateQuakingMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): QuakingMetrics => {
  if (!affixes.includes(14)) {
    return {
      damage: {},
      interrupts: {},
      hasQuaking: false,
    };
  }

  const damage = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageTaken" &&
      event.ability?.id === QUAKING &&
      event.targetPlayerID &&
      event.damage
    ) {
      acc[event.targetPlayerID] =
        (acc[event.targetPlayerID] ?? 0) + event.damage;
    }
    return acc;
  }, {});

  const interrupts = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "Interrupt" &&
      event.ability?.id === QUAKING &&
      event.sourcePlayerID
    ) {
      acc[event.sourcePlayerID] = (acc[event.sourcePlayerID] ?? 0) + 1;
    }
    return acc;
  }, {});

  return {
    damage,
    interrupts,
    hasQuaking: true,
  };
};

export type VolcanicMetrics = {
  hasVolcanic: boolean;
  damage: Record<number, number>;
  hits: Record<number, number>;
};

export const calculateVolcanicMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): VolcanicMetrics => {
  if (!affixes.includes(3)) {
    return {
      damage: {},
      hasVolcanic: false,
      hits: {},
    };
  }

  const { hits, damage } = events.reduce<{
    hits: Record<number, number>;
    damage: Record<number, number>;
  }>(
    (acc, event) => {
      if (
        event.type === "DamageTaken" &&
        event.ability?.id === VOLCANIC &&
        event.targetPlayerID
      ) {
        acc.hits[event.targetPlayerID] =
          (acc.hits[event.targetPlayerID] ?? 0) + 1;

        if (event.damage) {
          acc.damage[event.targetPlayerID] =
            (acc.damage[event.targetPlayerID] ?? 0) + event.damage;
        }
      }

      return acc;
    },
    {
      hits: {},
      damage: {},
    }
  );

  return {
    damage,
    hasVolcanic: true,
    hits,
  };
};

export type StormingMetrics = {
  hasStorming: boolean;
  damage: Record<number, number>;
  hits: Record<number, number>;
};

export const calculateStormingMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): StormingMetrics => {
  if (!affixes.includes(124)) {
    return {
      damage: {},
      hits: {},
      hasStorming: false,
    };
  }

  const { hits, damage } = events.reduce<{
    hits: Record<number, number>;
    damage: Record<number, number>;
  }>(
    (acc, event) => {
      if (
        event.type === "DamageTaken" &&
        event.ability?.id === STORMING &&
        event.targetPlayerID
      ) {
        acc.hits[event.targetPlayerID] =
          (acc.hits[event.targetPlayerID] ?? 0) + 1;

        if (event.damage) {
          acc.damage[event.targetPlayerID] =
            (acc.damage[event.targetPlayerID] ?? 0) + event.damage;
        }
      }

      return acc;
    },
    {
      hits: {},
      damage: {},
    }
  );

  return {
    damage,
    hits,
    hasStorming: true,
  };
};

export type NecroticMetrics = {
  hasNecrotic: boolean;
  damage: Record<number, number>;
};

export const calculateNecroticMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): NecroticMetrics => {
  if (!affixes.includes(4)) {
    return {
      damage: {},
      hasNecrotic: false,
    };
  }

  const damage = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageTaken" &&
      event.ability?.id === NECROTIC &&
      event.targetPlayerID &&
      event.damage
    ) {
      acc[event.targetPlayerID] =
        (acc[event.targetPlayerID] ?? 0) + event.damage;
    }
    return acc;
  }, {});

  return {
    hasNecrotic: true,
    damage,
  };
};

export type EncryptedMetrics = {
  hasEncrypted: boolean;
  damage: Record<number, number>;
};

export const calculateEncryptedMetrics = ({
  affixes,
  events,
}: CalculateAffixMetricsParams): EncryptedMetrics => {
  if (!affixes.includes(130)) {
    return {
      damage: {},
      hasEncrypted: false,
    };
  }

  const damage = events.reduce<Record<number, number>>((acc, event) => {
    if (
      event.type === "DamageTaken" &&
      event.ability &&
      ENCRYPTED.has(event.ability.id) &&
      event.targetPlayerID &&
      event.damage
    ) {
      acc[event.targetPlayerID] =
        (acc[event.targetPlayerID] ?? 0) + event.damage;
    }
    return acc;
  }, {});

  return {
    hasEncrypted: true,
    damage,
  };
};
