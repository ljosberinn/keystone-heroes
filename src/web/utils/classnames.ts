export const classnames = (...classes: unknown[]): string =>
  classes
    .filter((value) => typeof value !== "object" && Boolean(value))
    .join(" ");
