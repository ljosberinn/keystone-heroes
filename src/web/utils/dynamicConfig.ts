export const dynamicConfig = {
  suspense: true,
};
