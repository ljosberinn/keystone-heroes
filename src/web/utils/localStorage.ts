export const ls = (() => {
  const isSupported = (() => {
    if (typeof window === "undefined") {
      return false;
    }

    try {
      const key = "ls-test";
      localStorage.setItem(key, key);
      localStorage.removeItem(key);

      return true;
    } catch {
      return false;
    }
  })();

  return {
    set: (key: string, value: string) => {
      if (!isSupported) {
        return;
      }

      localStorage.setItem(key, value);
    },

    get: (key: string): string | null => {
      if (!isSupported) {
        return null;
      }

      return localStorage.getItem(key);
    },
    remove: (key: string) => {
      if (!isSupported) {
        return;
      }

      localStorage.removeItem(key);
    },
  };
})();
