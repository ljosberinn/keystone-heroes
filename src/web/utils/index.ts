import { isValidReportId } from "../../wcl/utils";

export const timeDurationToString = (time: number, omitMs = false): string => {
  const isNegative = time < 0;

  const inSeconds = time / 1000;
  const minutes = isNegative
    ? Math.ceil(inSeconds / 60)
    : Math.floor(inSeconds / 60);
  const seconds = isNegative
    ? Math.ceil(inSeconds - minutes * 60) * -1
    : Math.floor(inSeconds - minutes * 60);

  const prefix = inSeconds < 0 ? "-" : "";

  if (omitMs) {
    return `${prefix}${minutes}:${seconds.toString().padStart(2, "0")}`;
  }

  const ms =
    (time - minutes * 60 * 1000 - seconds * 1000) * (isNegative ? -1 : 1);

  return `${prefix}${minutes}:${seconds.toString().padStart(2, "0")}.${ms
    .toString()
    .slice(0, 3)
    .padStart(3, "0")}`;
};

type WCLUrlParams = {
  reportID: string;
  fightID: string;
  type?: "deaths" | "damage-done" | "healing" | "damage-taken";
  start?: number;
  end?: number;
  source?: number;
  ability?: number;
  hostility?: 1;
};

export const createWCLUrl = ({
  reportID: report,
  fightID: fight,
  ...rest
}: WCLUrlParams): string => {
  // @ts-expect-error fixing later
  const params = new URLSearchParams({
    ...rest,
    translate: "true",
  }).toString();

  const url = `https://www.warcraftlogs.com/reports/${report}#fight=${fight}`;

  return `${url}&${params}`;
};

type WowheadParams = {
  category: "spell" | "npc" | "affix";
  id: string | number;
};

export const createWowheadUrl = ({ category, id }: WowheadParams): string => {
  return `https://wowhead.com/${category}=${id}`;
};

export const parseWCLUrl = (
  maybeURL: string
): { reportID: null | string; fightID: null | string } => {
  const isReportID = isValidReportId(maybeURL);

  if (isReportID) {
    return {
      reportID: maybeURL,
      fightID: null,
    };
  }
  try {
    const { pathname, host, hash } = new URL(maybeURL);

    if (host === "www.warcraftlogs.com" && pathname.startsWith("/reports/")) {
      const maybeReportID = pathname.replace("/reports/", "").replace("/", "");

      if (!isValidReportId(maybeReportID)) {
        return {
          reportID: null,
          fightID: null,
        };
      }

      if (hash) {
        const maybeFightID = new URLSearchParams(hash.slice(1)).get("fight");

        if (
          maybeFightID &&
          (maybeFightID === "last" || Number.parseInt(maybeFightID) > 0)
        ) {
          return {
            reportID: maybeReportID,
            fightID: maybeFightID,
          };
        }
      }

      return {
        reportID: maybeReportID,
        fightID: null,
      };
    }

    return {
      reportID: null,
      fightID: null,
    };
  } catch {
    return {
      reportID: null,
      fightID: null,
    };
  }
};

export const classTextColorMap: Record<string, string> = {
  demonhunter: "text-demonhunter",
  warlock: "text-warlock",
  rogue: "text-rogue",
  warrior: "text-warrior",
  priest: "text-black dark:text-priest",
  hunter: "text-hunter",
  deathknight: "text-deathknight",
  shaman: "text-shaman",
  paladin: "text-paladin",
  monk: "text-monk",
  druid: "text-druid",
  mage: "text-mage",
};

export const classBorderColorMap: Record<string, string> = {
  demonhunter: "border-demonhunter",
  warlock: "border-warlock",
  rogue: "border-rogue",
  warrior: "border-warrior",
  priest: "border-priest",
  hunter: "border-hunter",
  deathknight: "border-deathknight",
  shaman: "border-shaman",
  paladin: "border-paladin",
  monk: "border-monk",
  druid: "border-druid",
  mage: "border-mage",
};
