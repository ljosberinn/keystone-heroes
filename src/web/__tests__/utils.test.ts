import { parseWCLUrl } from "../utils";

describe("parseWCLUrl", () => {
  test("fails given no url", () => {
    expect(parseWCLUrl("")).toMatchInlineSnapshot(`
      Object {
        "error": "INVALID_URL",
        "fightID": null,
        "reportID": null,
      }
    `);
  });

  test("fails given random url", () => {
    expect(parseWCLUrl("https://keystone-heroes.com")).toMatchInlineSnapshot(`
      Object {
        "error": "INVALID_HOST",
        "fightID": null,
        "reportID": null,
      }
    `);
  });

  test("fails given incomplete url", () => {
    expect(parseWCLUrl("https://keystone-heroes")).toMatchInlineSnapshot(`
      Object {
        "error": "INVALID_HOST",
        "fightID": null,
        "reportID": null,
      }
    `);
  });

  test("fails given invalid fight id", () => {
    expect(
      parseWCLUrl(
        "https://www.warcraftlogs.com/reports/aZ9y3jMctCqRvhKA#fight=kekw&type=damage-done"
      )
    ).toMatchInlineSnapshot(`
      Object {
        "error": null,
        "fightID": null,
        "reportID": "aZ9y3jMctCqRvhKA",
      }
    `);
  });

  test("passes given general report url", () => {
    expect(parseWCLUrl("https://www.warcraftlogs.com/reports/aZ9y3jMctCqRvhKA"))
      .toMatchInlineSnapshot(`
      Object {
        "error": null,
        "fightID": null,
        "reportID": "aZ9y3jMctCqRvhKA",
      }
    `);
  });

  test("passes given report url with fight id hash", () => {
    expect(
      parseWCLUrl(
        "https://www.warcraftlogs.com/reports/aZ9y3jMctCqRvhKA#fight=3"
      )
    ).toMatchInlineSnapshot(`
      Object {
        "error": null,
        "fightID": "3",
        "reportID": "aZ9y3jMctCqRvhKA",
      }
    `);
  });

  test("passes given report url with fight id === last", () => {
    expect(
      parseWCLUrl(
        "https://www.warcraftlogs.com/reports/aZ9y3jMctCqRvhKA#fight=last"
      )
    ).toMatchInlineSnapshot(`
      Object {
        "error": null,
        "fightID": "last",
        "reportID": "aZ9y3jMctCqRvhKA",
      }
    `);
  });

  test("passes given report url with trailing slash", () => {
    expect(
      parseWCLUrl(
        "https://www.warcraftlogs.com/reports/aZ9y3jMctCqRvhKA/#fight=last"
      )
    ).toMatchInlineSnapshot(`
      Object {
        "error": null,
        "fightID": "last",
        "reportID": "aZ9y3jMctCqRvhKA",
      }
    `);
  });
});
