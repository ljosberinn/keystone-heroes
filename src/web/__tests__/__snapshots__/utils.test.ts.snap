// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeDurationToString 1 negative 1`] = `"-00:00:001"`;

exports[`timeDurationToString 1 negative omitMs 1`] = `"-00:00"`;

exports[`timeDurationToString 1 negative toHours & omitMs 1`] = `"-01:00:00"`;

exports[`timeDurationToString 1 negative toHours 1`] = `"-01:00:00:001"`;

exports[`timeDurationToString 1 positive 1`] = `"00:00:001"`;

exports[`timeDurationToString 1 positive omitMs 1`] = `"00:00"`;

exports[`timeDurationToString 1 positive toHours & omitMs 1`] = `"01:00:00"`;

exports[`timeDurationToString 1 positive toHours 1`] = `"01:00:00:001"`;

exports[`timeDurationToString 10 negative 1`] = `"-00:00:010"`;

exports[`timeDurationToString 10 negative omitMs 1`] = `"-00:00"`;

exports[`timeDurationToString 10 negative toHours & omitMs 1`] = `"-01:00:00"`;

exports[`timeDurationToString 10 negative toHours 1`] = `"-01:00:00:010"`;

exports[`timeDurationToString 10 positive 1`] = `"00:00:010"`;

exports[`timeDurationToString 10 positive omitMs 1`] = `"00:00"`;

exports[`timeDurationToString 10 positive toHours & omitMs 1`] = `"01:00:00"`;

exports[`timeDurationToString 10 positive toHours 1`] = `"01:00:00:010"`;

exports[`timeDurationToString 100 negative 1`] = `"-00:00:100"`;

exports[`timeDurationToString 100 negative omitMs 1`] = `"-00:00"`;

exports[`timeDurationToString 100 negative toHours & omitMs 1`] = `"-01:00:00"`;

exports[`timeDurationToString 100 negative toHours 1`] = `"-01:00:00:100"`;

exports[`timeDurationToString 100 positive 1`] = `"00:00:100"`;

exports[`timeDurationToString 100 positive omitMs 1`] = `"00:00"`;

exports[`timeDurationToString 100 positive toHours & omitMs 1`] = `"01:00:00"`;

exports[`timeDurationToString 100 positive toHours 1`] = `"01:00:00:100"`;

exports[`timeDurationToString 1000 negative 1`] = `"-00:01:000"`;

exports[`timeDurationToString 1000 negative omitMs 1`] = `"-00:01"`;

exports[`timeDurationToString 1000 negative toHours & omitMs 1`] = `"-01:00:01"`;

exports[`timeDurationToString 1000 negative toHours 1`] = `"-01:00:01:000"`;

exports[`timeDurationToString 1000 positive 1`] = `"00:01:000"`;

exports[`timeDurationToString 1000 positive omitMs 1`] = `"00:01"`;

exports[`timeDurationToString 1000 positive toHours & omitMs 1`] = `"01:00:01"`;

exports[`timeDurationToString 1000 positive toHours 1`] = `"01:00:01:000"`;

exports[`timeDurationToString 10000 negative 1`] = `"-00:10:000"`;

exports[`timeDurationToString 10000 negative omitMs 1`] = `"-00:10"`;

exports[`timeDurationToString 10000 negative toHours & omitMs 1`] = `"-01:00:10"`;

exports[`timeDurationToString 10000 negative toHours 1`] = `"-01:00:10:000"`;

exports[`timeDurationToString 10000 positive 1`] = `"00:10:000"`;

exports[`timeDurationToString 10000 positive omitMs 1`] = `"00:10"`;

exports[`timeDurationToString 10000 positive toHours & omitMs 1`] = `"01:00:10"`;

exports[`timeDurationToString 10000 positive toHours 1`] = `"01:00:10:000"`;

exports[`timeDurationToString 100000 negative 1`] = `"-01:40:000"`;

exports[`timeDurationToString 100000 negative omitMs 1`] = `"-01:40"`;

exports[`timeDurationToString 100000 negative toHours & omitMs 1`] = `"-01:01:40"`;

exports[`timeDurationToString 100000 negative toHours 1`] = `"-01:01:40:000"`;

exports[`timeDurationToString 100000 positive 1`] = `"01:40:000"`;

exports[`timeDurationToString 100000 positive omitMs 1`] = `"01:40"`;

exports[`timeDurationToString 100000 positive toHours & omitMs 1`] = `"01:01:40"`;

exports[`timeDurationToString 100000 positive toHours 1`] = `"01:01:40:000"`;

exports[`timeDurationToString 1000000 negative 1`] = `"-16:40:000"`;

exports[`timeDurationToString 1000000 negative omitMs 1`] = `"-16:40"`;

exports[`timeDurationToString 1000000 negative toHours & omitMs 1`] = `"-01:16:40"`;

exports[`timeDurationToString 1000000 negative toHours 1`] = `"-01:16:40:000"`;

exports[`timeDurationToString 1000000 positive 1`] = `"16:40:000"`;

exports[`timeDurationToString 1000000 positive omitMs 1`] = `"16:40"`;

exports[`timeDurationToString 1000000 positive toHours & omitMs 1`] = `"01:16:40"`;

exports[`timeDurationToString 1000000 positive toHours 1`] = `"01:16:40:000"`;

exports[`timeDurationToString 10000000 negative 1`] = `"-46:40:000"`;

exports[`timeDurationToString 10000000 negative omitMs 1`] = `"-46:40"`;

exports[`timeDurationToString 10000000 negative toHours & omitMs 1`] = `"-03:46:40"`;

exports[`timeDurationToString 10000000 negative toHours 1`] = `"-03:46:40:000"`;

exports[`timeDurationToString 10000000 positive 1`] = `"46:40:000"`;

exports[`timeDurationToString 10000000 positive omitMs 1`] = `"46:40"`;

exports[`timeDurationToString 10000000 positive toHours & omitMs 1`] = `"03:46:40"`;

exports[`timeDurationToString 10000000 positive toHours 1`] = `"03:46:40:000"`;

exports[`timeDurationToString 100000000 negative 1`] = `"-46:40:000"`;

exports[`timeDurationToString 100000000 negative omitMs 1`] = `"-46:40"`;

exports[`timeDurationToString 100000000 negative toHours & omitMs 1`] = `"-04:46:40"`;

exports[`timeDurationToString 100000000 negative toHours 1`] = `"-04:46:40:000"`;

exports[`timeDurationToString 100000000 positive 1`] = `"46:40:000"`;

exports[`timeDurationToString 100000000 positive omitMs 1`] = `"46:40"`;

exports[`timeDurationToString 100000000 positive toHours & omitMs 1`] = `"04:46:40"`;

exports[`timeDurationToString 100000000 positive toHours 1`] = `"04:46:40:000"`;

exports[`timeDurationToString 1000000000 negative 1`] = `"-46:40:000"`;

exports[`timeDurationToString 1000000000 negative omitMs 1`] = `"-46:40"`;

exports[`timeDurationToString 1000000000 negative toHours & omitMs 1`] = `"-14:46:40"`;

exports[`timeDurationToString 1000000000 negative toHours 1`] = `"-14:46:40:000"`;

exports[`timeDurationToString 1000000000 positive 1`] = `"46:40:000"`;

exports[`timeDurationToString 1000000000 positive omitMs 1`] = `"46:40"`;

exports[`timeDurationToString 1000000000 positive toHours & omitMs 1`] = `"14:46:40"`;

exports[`timeDurationToString 1000000000 positive toHours 1`] = `"14:46:40:000"`;

exports[`timeDurationToString 10000000000 negative 1`] = `"-46:40:000"`;

exports[`timeDurationToString 10000000000 negative omitMs 1`] = `"-46:40"`;

exports[`timeDurationToString 10000000000 negative toHours & omitMs 1`] = `"-18:46:40"`;

exports[`timeDurationToString 10000000000 negative toHours 1`] = `"-18:46:40:000"`;

exports[`timeDurationToString 10000000000 positive 1`] = `"46:40:000"`;

exports[`timeDurationToString 10000000000 positive omitMs 1`] = `"46:40"`;

exports[`timeDurationToString 10000000000 positive toHours & omitMs 1`] = `"18:46:40"`;

exports[`timeDurationToString 10000000000 positive toHours 1`] = `"18:46:40:000"`;
